## Preprocessing

In [159]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import sqlite3

# Read database from 'Data/bank_additional.sqlite'.

conn = sqlite3.connect('Data/bank_additional.sqlite')
df = pd.read_sql("select * from marketing",conn)

# Close database
conn.close()

df.head()

age          job  marital          education default  housing     loan  \
0   30  blue-collar  married           basic.9y      no      yes       no   
1   39     services   single        high.school      no       no       no   
2   25     services  married        high.school      no      yes       no   
3   38     services  married           basic.9y      no  unknown  unknown   
4   47       admin.  married  university.degree      no      yes       no   

     contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0   cellular   may         fri  ...         2    999         0  nonexistent   
1  telephone   may         fri  ...         4    999         0  nonexistent   
2  telephone   jun         wed  ...         1    999         0  nonexistent   
3  telephone   jun         fri  ...         3    999         0  nonexistent   
4   cellular   nov         mon  ...         1    999         0  nonexistent   

  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0         -1.8          92.893          -46.2      1.313       5099.1  no  
1          1.1          93.994          -36.4      4.855       5191.0  no  
2          1.4          94.465          -41.8      4.962       5228.1  no  
3          1.4          94.465          -41.8      4.959       5228.1  no  
4         -0.1          93.200          -42.0      4.191       5195.8  no  

[5 rows x 21 columns]

In [160]:
df=df.dropna()
df

age          job  marital          education default  housing     loan  \
0      30  blue-collar  married           basic.9y      no      yes       no   
1      39     services   single        high.school      no       no       no   
2      25     services  married        high.school      no      yes       no   
3      38     services  married           basic.9y      no  unknown  unknown   
4      47       admin.  married  university.degree      no      yes       no   
...   ...          ...      ...                ...     ...      ...      ...   
4114   30       admin.  married           basic.6y      no      yes      yes   
4115   39       admin.  married        high.school      no      yes       no   
4116   27      student   single        high.school      no       no       no   
4117   58       admin.  married        high.school      no       no       no   
4118   34   management   single        high.school      no      yes       no   

        contact month day_of_week  ...  campaign  pdays  previous  \
0      cellular   may         fri  ...         2    999         0   
1     telephone   may         fri  ...         4    999         0   
2     telephone   jun         wed  ...         1    999         0   
3     telephone   jun         fri  ...         3    999         0   
4      cellular   nov         mon  ...         1    999         0   
...         ...   ...         ...  ...       ...    ...       ...   
4114   cellular   jul         thu  ...         1    999         0   
4115  telephone   jul         fri  ...         1    999         0   
4116   cellular   may         mon  ...         2    999         1   
4117   cellular   aug         fri  ...         1    999         0   
4118   cellular   nov         wed  ...         1    999         0   

         poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0     nonexistent         -1.8          92.893          -46.2      1.313   
1     nonexistent          1.1          93.994          -36.4      4.855   
2     nonexistent          1.4          94.465          -41.8      4.962   
3     nonexistent          1.4          94.465          -41.8      4.959   
4     nonexistent         -0.1          93.200          -42.0      4.191   
...           ...          ...             ...            ...        ...   
4114  nonexistent          1.4          93.918          -42.7      4.958   
4115  nonexistent          1.4          93.918          -42.7      4.959   
4116      failure         -1.8          92.893          -46.2      1.354   
4117  nonexistent          1.4          93.444          -36.1      4.966   
4118  nonexistent         -0.1          93.200          -42.0      4.120   

      nr.employed   y  
0          5099.1  no  
1          5191.0  no  
2          5228.1  no  
3          5228.1  no  
4          5195.8  no  
...           ...  ..  
4114       5228.1  no  
4115       5228.1  no  
4116       5099.1  no  
4117       5228.1  no  
4118       5195.8  no  

[4119 rows x 21 columns]

In [161]:
# Determine the number of unique values in each column.
df.nunique()

age                67
job                12
marital             4
education           8
default             3
housing             3
loan                3
contact             2
month              10
day_of_week         5
duration          828
campaign           25
pdays              21
previous            7
poutcome            3
emp.var.rate       10
cons.price.idx     26
cons.conf.idx      26
euribor3m         234
nr.employed        11
y                   2
dtype: int64

In [162]:
# Convert categorical data to numeric with `pd.get_dummies`
df = pd.get_dummies(df,dtype=float)
df.head()

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0   30       487         2    999         0          -1.8          92.893   
1   39       346         4    999         0           1.1          93.994   
2   25       227         1    999         0           1.4          94.465   
3   38        17         3    999         0           1.4          94.465   
4   47        58         1    999         0          -0.1          93.200   

   cons.conf.idx  euribor3m  nr.employed  ...  day_of_week_fri  \
0          -46.2      1.313       5099.1  ...              1.0   
1          -36.4      4.855       5191.0  ...              1.0   
2          -41.8      4.962       5228.1  ...              0.0   
3          -41.8      4.959       5228.1  ...              1.0   
4          -42.0      4.191       5195.8  ...              0.0   

   day_of_week_mon  day_of_week_thu  day_of_week_tue  day_of_week_wed  \
0              0.0              0.0              0.0              0.0   
1              0.0              0.0              0.0              0.0   
2              0.0              0.0              0.0              1.0   
3              0.0              0.0              0.0              0.0   
4              1.0              0.0              0.0              0.0   

   poutcome_failure  poutcome_nonexistent  poutcome_success  y_no  y_yes  
0               0.0                   1.0               0.0   1.0    0.0  
1               0.0                   1.0               0.0   1.0    0.0  
2               0.0                   1.0               0.0   1.0    0.0  
3               0.0                   1.0               0.0   1.0    0.0  
4               0.0                   1.0               0.0   1.0    0.0  

[5 rows x 65 columns]

In [163]:
# Split our preprocessed data into our features and target arrays
y = df['y_yes'].values
X = df.drop(['y_yes','y_no'], axis=1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state = 42)

In [164]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [165]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 = 3
hidden_nodes_layer2 = 6
hidden_nodes_layer3 = 12

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="tanh")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="tanh"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="tanh"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_20"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_74 (Dense)            (None, 3)                 192       
                                                                 
 dense_75 (Dense)            (None, 6)                 24        
                                                                 
 dense_76 (Dense)            (None, 12)                84        
                                                                 
 dense_77 (Dense)            (None, 1)                 13        
                                                                 
Total params: 313
Trainable params: 313
Non-trainable params: 0
_________________________________________________________________


In [166]:
# Compile the model
nn.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])

In [167]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train, epochs=100)

Epoch 1/100
97/97 [==============================] - 5s 12ms/step - loss: 0.5971 - accuracy: 0.8019
Epoch 2/100
97/97 [==============================] - 1s 7ms/step - loss: 0.4053 - accuracy: 0.8893
Epoch 3/100
97/97 [==============================] - 1s 5ms/step - loss: 0.2952 - accuracy: 0.8938
Epoch 4/100
97/97 [==============================] - 0s 4ms/step - loss: 0.2479 - accuracy: 0.8990
Epoch 5/100
97/97 [==============================] - 0s 4ms/step - loss: 0.2232 - accuracy: 0.9055
Epoch 6/100
97/97 [==============================] - 0s 4ms/step - loss: 0.2075 - accuracy: 0.9094
Epoch 7/100
97/97 [==============================] - 0s 4ms/step - loss: 0.1983 - accuracy: 0.9168
Epoch 8/100
97/97 [==============================] - 0s 4ms/step - loss: 0.1935 - accuracy: 0.9171
Epoch 9/100
97/97 [==============================] - 0s 4ms/step - loss: 0.1891 - accuracy: 0.9187
Epoch 10/100
97/97 [==============================] - 0s 5ms/step - loss: 0.1858 - accuracy: 0.9187
Epoch 11

## Optimisation
Firstly, we decreased the first hidden layer neurons from 7 to 3, the second hidden layer neurons from 14 to 6.

Secondly, we added the third hidden layer neurons.

Thirdly, we used different activation function for the hidden layer, from ReLU to Tanh.

Finally, Tanh is better than ReLu, and three hidden layers better than two. the best accuracy was composed of Tanh and 3/6/12 layer

### Function **ReLU**
Layer 1 / Layer 2 / Layer 3  
+ 7/14:  
33/33 - 1s - loss: 0.3537 - accuracy: 0.8835 - 844ms/epoch - 26ms/step  
Loss: 0.35365232825279236, Accuracy: 0.8834951519966125
+ 7/14/21:  
33/33 - 1s - loss: 0.7113 - accuracy: 0.8854 - 505ms/epoch - 15ms/step  
Loss: 0.7113155722618103, Accuracy: 0.8854368925094604
+ 3/6:  
33/33 - 1s - loss: 0.3223 - accuracy: 0.8961 - 934ms/epoch - 28ms/step  
Loss: 0.3223118185997009, Accuracy: 0.8961164951324463
+ 3/6/12:  
33/33 - 1s - loss: 0.2780 - accuracy: 0.8786 - 842ms/epoch - 26ms/step  
Loss: 0.2780188024044037, Accuracy: 0.8786407709121704
### Function **Tanh**
Layer 1 / Layer 2 / Layer 3  
+ 3/6:  
33/33 - 1s - loss: 0.2194 - accuracy: 0.8990 - 753ms/epoch - 23ms/step  
Loss: 0.2194284200668335, Accuracy: 0.8990291357040405
+ 3/6/12:  
33/33 - 0s - **loss: 0.2193 - accuracy: 0.9049** - 294ms/epoch - 9ms/step  
Loss: 0.21933530271053314, Accuracy: 0.9048543572425842
+ 7/14/21:  
33/33 - 1s - loss: 0.6689 - accuracy: 0.9000 - 895ms/epoch - 27ms/step  
Loss: 0.6688849925994873, Accuracy: 0.8999999761581421

In [168]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

33/33 - 0s - loss: 0.2193 - accuracy: 0.9049 - 294ms/epoch - 9ms/step
Loss: 0.21933530271053314, Accuracy: 0.9048543572425842


In [169]:
# Export our model to HDF5 file
nn.save("Bank_Marketing.h5")